File: ObjectiveC/FlyingLogo/ConfigureSheet/ConfigureSheetController.m
//
//  ConfigureSheetController.m
//  FlyingLogo
//
//  Created by Gergely Sánta on 25/09/2019.
//  Copyright © 2019 TriKatz. All rights reserved.
//

#import "ConfigureSheetController.h"
#import "Configuration.h"

@interface ConfigureSheetController ()

@property (strong) IBOutlet NSPopUpButton *sceneTypeSelector;

@end

@implementation ConfigureSheetController

- (NSNibName) windowNibName {
	return @"ConfigureSheet";
}

- (void) windowDidLoad {
	[super windowDidLoad];

	// Create new menu for sceneTypeSelector
	NSMenu *sceneTypesMenu = [NSMenu new];

	// Fill up menu with available scene names
	for (NSString *name in Configuration.shared.sceneNames) {
		NSMenuItem *item = [NSMenuItem new];
		item.title = name;
		item.action = @selector(sceneTypeSelected:);
		item.tag = [(NSNumber*)[Configuration.shared.scenes valueForKey:name] intValue];
		[sceneTypesMenu addItem:item];
	}

	// Set new menu for selector and select actually configured value
	self.sceneTypeSelector.menu = sceneTypesMenu;
	[self.sceneTypeSelector selectItemWithTag:Configuration.shared.sceneType];
}

- (void) sceneTypeSelected:(NSMenuItem*)menuItem {
	// Store selection
	Configuration.shared.sceneType = menuItem.tag;
}

- (IBAction) closeButtonPressed:(NSButton *)sender {
	[NSApp endSheet:self.window];
}

@end

File: Swift/FlyingLogo/ConfigureSheet/ConfigureSheetController.swift
//
//  ConfigureSheetController.swift
//  FlyingLogo
//
//  Created by Gergely Sánta on 28/09/2019.
//  Copyright © 2019 TriKatz. All rights reserved.
//

import Cocoa

class ConfigureSheetController: NSWindowController {

	@IBOutlet var sceneTypeSelector: NSPopUpButton!
	
	override var windowNibName: NSNib.Name? {
		return "ConfigureSheet"
	}

	override func windowDidLoad() {
		super.windowDidLoad()

		// Create new menu for sceneTypeSelector
		let sceneTypesMenu = NSMenu()

		// Fill up menu with available scene names
		for name in Configuration.shared.sceneNames {
			let item = NSMenuItem()
			item.title = name
			item.action = #selector(sceneTypeSelected(menuItem:))
			item.tag = (Configuration.shared.scenes[name] ?? Configuration.SceneType.spaceFly).rawValue
			sceneTypesMenu.addItem(item)
		}

		// Set new menu for selector and select actually configured value
		self.sceneTypeSelector.menu = sceneTypesMenu
		self.sceneTypeSelector.selectItem(withTag: Configuration.shared.sceneType.rawValue)
	}

	@objc func sceneTypeSelected(menuItem: NSMenuItem) {
		// Store selection
		Configuration.shared.sceneType = Configuration.SceneType(rawValue: menuItem.tag) ?? Configuration.SceneType.spaceFly
	}

	@IBAction func closeButtonPressed(_ sender: NSButton) {
		if let window = self.window {
			NSApp.endSheet(window)
		}
	}
	
}
